dyninstAPI/src/inst-aarch64.C

```diff
@@ -1389,37 +1422,257 @@ bool EmitterAARCH64::emitCallInstruction(codeGen &gen, func_instance *callee, bo
-    assert(0); //Not implemented
+    //assert(0); //Not implemented
+    //return;
+    
+    // create or retrieve jump slot
+    Address addr;
+    int stackSize = 0;
+
+    if(var == NULL) {
+        addr = offset;
+    }
+    else if(!is_local) {
+        addr = getInterModuleVarAddr(var, gen);
+    }
+    else {
+        addr = (Address)var->getOffset();
+    }
+
+    // load register with address from jump slot
+
+    printf("emitLoadShared addr 0x%lx curr adress 0x%lx offset %ld 0x%lx size %d\n", 
+            addr, gen.currAddr(), addr - gen.currAddr()+4, addr - gen.currAddr()+4, size);
+    Register scratchReg = gen.rs()->getScratchRegister(gen, true);
+
+    if (scratchReg == REG_NULL) {
+        assert(0);
+    }
+
+    emitMovePCToReg(scratchReg, gen);
+    Address varOffset = addr - gen.currAddr()+4;
+
+    if (op ==loadOp) {
+        if(!is_local && (var != NULL)){
+            emitLoadRelative(dest, varOffset, scratchReg, gen.width(), gen);
+            // Deference the pointer to get the variable
+            // emitLoadRelative(dest, 0, dest, size, gen);
+            // Offset mode to load back to itself
+            printf("Fallback to offset mode\n");
+            insnCodeGen::generateMemAccess(gen, insnCodeGen::Load, dest, dest, 0, 8, insnCodeGen::Offset);
+        } else {
+            emitLoadRelative(dest, varOffset, scratchReg, size, gen);
+        }
+    } else { //loadConstop
+        if(!is_local && (var != NULL)){
+            emitLoadRelative(dest, varOffset, scratchReg, gen.width(), gen);
+        } else {
+            printf("oops, reached else branch\n");
+            assert(0);
+
+            // Move address of the variable into the register - load effective address
+            //dest = effective address of pc+offset ;
+            //insnCodeGen::generateImm (gen, CAUop, dest, 0, BOT_HI (varOffset));
+            //insnCodeGen::generateImm (gen, ORILop, dest, dest, BOT_LO (varOffset));
+            //insnCodeGen::generateAddReg (gen, CAXop, dest, dest, scratchReg);
+        }
+    }
+
+    if (stackSize > 0)
+        assert(0);
+
     return;
 }
 
 void
 EmitterAARCH64::emitStoreShared(Register source, const image_variable *var, bool is_local, int size, codeGen &gen) {
-    assert(0); //Not implemented
+    //assert(0); //Not implemented
+    //return;
+    //
+    // l: from POWER
+    //
+    // create or retrieve jump slot
+    Address addr;
+    int stackSize = 0;
+    if(!is_local) {
+        addr = getInterModuleVarAddr(var, gen);
+    }
+    else {
+        addr = (Address)var->getOffset();
+    }
+
+    printf("emitStoreRelative addr 0x%lx curr adress 0x%lx offset %ld 0x%lx size %d\n",
+            addr, gen.currAddr(), addr - gen.currAddr()+4, addr - gen.currAddr()+4, size);
+
+    // load register with address from jump slot
+    Register scratchReg = gen.rs()->getScratchRegister(gen, true);
+    if (scratchReg == REG_NULL) {
+        assert(0);
```

Comment:
Please add useful error message here- cf. line 1557.

Reply:
I've added details for asserts, and used "assert(0 && "msg")" as most of the code does, hope that's ok.